### PR TITLE
Reduce object creation for for unwrap/wrap if no ByteBuffer[] is used.

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

Often unwrap(...), wrap(...) is used with a single ByteBuffer and not with a ByteBuffer[]. We should reduce the array creations in this case.

Modifications:

Reuse ByteBuffer[1] for dst/src ByteBuffer.

Result:

Less object creation and so less GC

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler @trustin PTAL

File: handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java [hunk 1]
```diff
@@ -786,10 +788,60 @@ public SSLEngineResult unwrap(final ByteBuffer[] srcs, final ByteBuffer[] dsts)
...
+        return singleDstBuffer;
+    }
+
+    private void resetSingleDstBuffer() {
+        singleDstBuffer[0] = null;
+    }
+
     @Override
-    public SSLEngineResult unwrap(
+    public synchronized SSLEngineResult unwrap(
```
Comment (nmittler) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java]: What sort of impact do we expect from adding these synchronization blocks?

Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java]: @nmittler it does not matter as the super.unwrap(...) method just delegated to our already synchronized unwrap method.

Comment (nmittler) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java]: @normanmaurer ah ok ... makes sense.  Thanks.

**Comment by nmittler** [comment]
@normanmaurer ship it! :)

**Comment by Scottmitch** [comment]
@normanmaurer - Ship it after build failure addressed.

**Comment by normanmaurer** [comment]
Build failure is not related

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (80de5fa9c9c8e0477d5f6679622c529f3ddf9e6d), 4.1 (66c3c58d3e52c22d32a1b41bdbb6b9ab47bdc2fe) and master (207d0ab2ab6471071d8bb12672d897dda7223299)

**Pull request closed without merging**